src/actions/ensRegistryActions.js

```diff
@@ -0,0 +1,95 @@
+// @flow
+/*
+    Pillar Wallet: the personal data locker
+    Copyright (C) 2019 Stiftung Pillar Project
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with this program; if not, write to the Free Software Foundation, Inc.,
+    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+import get from 'lodash.get';
+
+// constants
+import { ADD_ENS_REGISTRY_RECORD, SET_ENS_REGISTRY_RECORDS } from 'constants/ensRegistryConstants';
+
+// models, types
+import type { Dispatch, GetState } from 'reducers/rootReducer';
+
+// utils
+import { lookupAddress } from 'utils/common';
+
+// actions
+import { saveDbAction } from './dbActions';
+
+
+export const addEnsRegistryRecordAction = (address: string, ensName: string) => {
+  return async (dispatch: Dispatch, getState: GetState) => {
+    const { ensRegistry: { data: ensRegistry } } = getState();
+    if (ensRegistry[address]) return;
+
+    dispatch(saveDbAction('ensRegistry', { ensRegistry: { [address]: ensName } }));
+    dispatch({
+      type: ADD_ENS_REGISTRY_RECORD,
+      payload: {
+        address,
+        ensName,
+      },
+    });
+  };
+};
+
+export const lookupAddressAction = (address: string) => {
+  return async (dispatch: Dispatch, getState: GetState) => {
+    const { ensRegistry: { data: ensRegistry } } = getState();
+    if (ensRegistry[address]) return;
+
+    const ensName = await lookupAddress(address);
+    if (ensName) {
+      dispatch(addEnsRegistryRecordAction(address, ensName));
+    }
+  };
+};
+
+export const extractEnsInfoFromTransactionsAction = (transactions: Object[]) => {
+  return async (dispatch: Dispatch, getState: GetState) => {
+    const { ensRegistry: { data: ensRegistry } } = getState();
+    let dataUpdated = false;
+
+    transactions.forEach(transaction => {
+      const fromEnsName = get(transaction, 'from.account.ensName') || get(transaction, 'sender.account.ensName');
+      const fromAddress = get(transaction, 'from.account.address') || get(transaction, 'sender.account.address');
+      const toEnsName = get(transaction, 'to.account.ensName') || get(transaction, 'recipient.account.ensName');
+      const toAddress = get(transaction, 'to.account.address') || get(transaction, 'recipient.account.address');
```

Comment:
What about something like:

```js
const sender = transaction.from || transaction.sender;
const recipient = transaction.to || transaction.recipient;
const {
  ensName: fromEnsName,
  address: fromAddress,
} = sender.account; // I think this is safe, but we could do: get(sender, 'account', {});
const {
  ensName: toEnsName,
  address: toAddress,
} = recipient.account; // I think this is also safe, or we could do: get(recipient, 'account', {});
```

I'm assuming we're always gonna get everything under either `transaction.from.account` or `transaction.sender.account`. If we receive data in both, it won't work. For example:

```json
{
  "transaction": {
    "from": {
      "account": {
        "ensName": "name"
      }
    },
    "sender": {
      "account": {
        "address": "0x..."
      }
    }
}
```

Reply:
Unfortunately, the `account` property is optional, since you can get a transaction from non-registered address. In that case, as I remember, `from.account` will be null, and you get the `from.address` property as a string (without the ens value).

We could change this block to that one:
```javascript
      const sender = transaction.from || transaction.sender;
      const recipient = transaction.to || transaction.recipient;

      const fromEnsName = get(sender, 'account.ensName');
      const fromAddress = get(sender, 'account.address');
      const toEnsName = get(recipient, 'account.ensName');
      const toAddress = get(recipient, 'account.address');
```
what do you think?